driver-core/src/test/functional/com/mongodb/client/model/expressions/DocumentExpressionsFunctionalTest.java

```diff
@@ -0,0 +1,238 @@
+        assertExpression(2007, ofDoc("{a: {'$date': '2007-12-03T10:15:30.005Z'}}")
+                .getDate("a", Instant.EPOCH).year(of("UTC")));
+        // no convenience for arrays
+        assertExpression(Document.parse("{b: 2}"), ofDoc("{a: {b: 2}}")
+                .getDocument("a", Document.parse("{z: 99}")));
+
+        // normal
+        assertExpression(true, ofDoc("{a: true}").getBoolean("a", of(false)));
+        assertExpression(1.0, ofDoc("{a: 1.0}").getNumber("a", of(99)));
+        assertExpression("A", ofDoc("{a: 'A'}").getString("a", of("Z")));
+        assertExpression(2007, ofDoc("{a: {'$date': '2007-12-03T10:15:30.005Z'}}")
+                .getDate("a", of(Instant.EPOCH)).year(of("UTC")));
+        assertExpression(Arrays.asList(3, 2), ofDoc("{a: [3, 2]}").getArray("a", ofIntegerArray(99, 88)));
+        assertExpression(Document.parse("{b: 2}"), ofDoc("{a: {b: 2}}")
+                .getDocument("a", of(Document.parse("{z: 99}"))));
+
+        // right branch (missing field)
+        assertExpression(false, ofDoc("{}").getBoolean("a", false));
+        assertExpression(99, ofDoc("{}").getNumber("a", 99));
+        assertExpression(99L, ofDoc("{}").getNumber("a", 99L));
+        assertExpression(99.0, ofDoc("{}").getNumber("a", 99.0));
+        assertExpression(Decimal128.parse("99"), ofDoc("{}").getNumber("a", Decimal128.parse("99")));
+        assertExpression("Z", ofDoc("{}").getString("a", "Z"));
+        assertExpression(1970, ofDoc("{}")
+                .getDate("a", Instant.EPOCH).year(of("UTC")));
+        assertExpression(Arrays.asList(99, 88), ofDoc("{}").getArray("a", ofIntegerArray(99, 88)));
+        assertExpression(Document.parse("{z: 99}"), ofDoc("{}")
+                .getDocument("a", Document.parse("{z: 99}")));
+    }
+
+    @Test
+    public void getFieldMissingTest() {
+        // missing fields
+        assertExpression(
+                BsonDocument.parse("{'a': 1}"),
+                a1.setField("z", a1.getBoolean("missing")));
+        assertExpression(
+                BsonDocument.parse("{'a': 1}"),
+                a1.setField("z", a1.getDocument("missing").getDocument("also_missing")));
+        assertExpression(
+                BsonDocument.parse("{'a': 1, 'z': ''}"),
+                a1.setField("z", a1.getString("missing").toLower()));
+        /*
+        The behaviour of missing fields appears to be as follows, and equivalent to $$REMOVE:
+        propagates -- getField, cond branches...
+        false -- not, or, cond check...
+        0     -- sum...
+        ""    -- toLower...
+        null  -- multiply, add, subtract, year, filter, reduce, map, result within map...
+        */
+    }
+
+    @Test
+    public void setFieldTest() {
+        // https://www.mongodb.com/docs/manual/reference/operator/aggregation/setField/
+        // "setField" is "withField": nothing is modified, and a new copy is produced.
+        // Placing a field based on a literal:
```

Comment:
Comments like this one, or the one above appear to be a replacement for the API docs that are not being written at the moment. I commented on similar comments in https://github.com/mongodb/mongo-java-driver/pull/1050. What will happen to them eventually? I am not sure they should be left in the test code.

Reply:
165 is a clarification that belongs on the javadocs, but also doesn't hurt here in case the test starts failing. 166 is just a normal doc, which in this method I have placed above each test because they get fairly long and confusing.

Reply:
I would generally read comments in failing tests.

Certain comments explain why certain things are or are not tested, or why tests are written or behave in a certain way. I am not sure what criteria to use for including or excluding such comments from tests, so I cannot hope to reliably exclude them in the future, though I don't have strong objections to removing them if some objection is raised in a PR.